File: leetcode/Arrays/leetcode-215/215.go
package arrays

func findKthLargest(nums []int, k int) int {
	left, right := 0, len(nums)-1
	for {
		pos := partition(nums, left, right)
		if pos == k-1 {
			return pos
		} else if pos > k-1 {
			right = pos - 1
		} else {
			left = pos + 1
		}
	}
}

func partition(nums []int, left, right int) int {
	pos := left
	pivot := nums[right]
	i := left

	for i < right {
		if nums[i] > pivot {
			nums[i], nums[pos] = nums[pos], nums[i]
			pos++
		}
		i++
	}
	nums[pos], nums[right] = nums[right], nums[pos]
	return pos
}

File: leetcode/Arrays/leetcode-215/215.java
class Solution {
    public int findKthLargest(int[] nums, int k) {
        int left = 0, right = nums.length - 1;
        while (true) {
            int pos = partition(nums, left, right);
            if (pos == k - 1) {
                return nums[pos];
            } else if (pos > k - 1) {
                right = pos - 1;
            } else {
                left = pos + 1;
            }
        }
    }

    private int partition(int[] nums, int left, int right) {
        int pivot = nums[right];
        int pos = left;
        int i = left;
        while (i < right) {
            if (nums[i] < pivot) {
                swap(nums, pos, i);
                pos++;
            }
            i++;
        }
        swap(nums, pos, right);
        return pos;
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}